### PR TITLE
US132742 - Add/Edit Schedule - Front-end validation

### PR DESCRIPTION
**Pull request by jguann**
[US132742](https://rally1.rallydev.com/#/431372673576ud/iterationstatus?detail=%2Fuserstory%2F611411896311)

- Validate required fields exist (name, subject OR catalog & section, delivery blocks)
- Validate certain fields are csv format if they exist (subject)
- Handle back-end saving error with generic JSON message

image
image
image

### CODE REVIEW

File: src/components/module-scheduler-schedule-dialog.js [hunk 1]
```diff
@@ -188,8 +214,30 @@ class ScheduleDialog extends LocalizeMixin(LitElement) {
...
+		if (!(this.sessionCode || this.subjectCode)) {
+			this._missingSubjectOrSessionField = true;
+		}
+
+		// Check alphanumeric comma separated value format
+		const csvPattern = '^[a-zA-Z0-9]+(,\\s*[a-zA-Z0-9]+)*$';
+
+		if (this.subjectCode && !this.subjectCode.match(csvPattern)) {
+			this._invalidSubjectField = true;
+		}
```
Comment (Syeda-Azim) [inline src/components/module-scheduler-schedule-dialog.js:233]: Should we trim subjectCode before matching csv pattern to skip blank space ?

File: src/components/module-scheduler-schedule-dialog.js [hunk 2]
```diff
@@ -294,17 +339,21 @@ class ScheduleDialog extends LocalizeMixin(LitElement) {
 			}
 			this.closeDialog(SAVE_ACTION);
 		} catch (err) {
-			// TODO: handle errors
+			this._invalidFormAlertMessage = 'Please check your JSON and try again.';
```
Comment (Syeda-Azim) [inline src/components/module-scheduler-schedule-dialog.js:342]: Error message should be loaded from langterms.

File: src/components/module-scheduler-schedule-dialog.js [hunk 3]
```diff
@@ -317,12 +366,73 @@ class ScheduleDialog extends LocalizeMixin(LitElement) {
...
+		this._missingSubjectOrSessionField = false;
+		this._invalidSubjectField = false;
 		this.subjectCode = e.target.value;
 	}
+
+	_renderJsonErrorTooltip() {
+		if (this._invalidJsonField) {
+			return html`
+			<d2l-tooltip for="scheduleJson" state="error" align="start" offset="10">
+					Delivery Blocks must be JSON with valid dates.
```
Comment (Syeda-Azim) [inline src/components/module-scheduler-schedule-dialog.js:383]: Same here.

File: src/components/module-scheduler-schedule-dialog.js [hunk 4]
```diff
@@ -317,12 +366,73 @@ class ScheduleDialog extends LocalizeMixin(LitElement) {
...
+			</d2l-tooltip>
+			`;
+		}
+	}
+
+	_renderModuleIgnoreStringsErrorTooltip() {
+		if (this._invalidModuleIgnoreStringsField) {
+			return html`
+			<d2l-tooltip for="moduleIgnoreList" state="error" align="start" offset="10">
+					Module Ignore Strings must be a comma separated value.
```
Comment (Syeda-Azim) [inline src/components/module-scheduler-schedule-dialog.js:393]: Same here.

File: src/components/module-scheduler-schedule-dialog.js (hunk 1 above)
Comment (jguann) [reply src/components/module-scheduler-schedule-dialog.js:233]: Ah yes, the pattern won't match if there are leading or trailing spaces so I'll trim it of all whitespaces.

File: src/components/module-scheduler-schedule-dialog.js [hunk 5]
```diff
@@ -226,14 +225,20 @@ class ScheduleDialog extends LocalizeMixin(LitElement) {
...
+			this.subjectCode = this.subjectCode.split(' ').join('');
+			if (!this.subjectCode.match(csvPattern)) {
+				this._invalidSubjectField = true;
+			}
 		}
 
-		if (this.moduleIgnoreList && !this.moduleIgnoreList.match(csvPattern)) {
-			this._invalidModuleIgnoreStringsField = true;
+		if (this.moduleIgnoreList) {
+			this.moduleIgnoreList = this.moduleIgnoreList.split(' ').join('');
```
Comment (Syeda-Azim) [inline src/components/module-scheduler-schedule-dialog.js:234]: This can be done simply by `trim()` .

Comment (jguann) [reply src/components/module-scheduler-schedule-dialog.js:234]: I wanted to remove all the whitespaces within the string, not just leading and trailing.

Comment (IhabAlkabary) [reply src/components/module-scheduler-schedule-dialog.js:234]: module names could have spaces within, so we shouldn't do that

Comment (jguann) [reply src/components/module-scheduler-schedule-dialog.js:234]: So spaces in between alphanumeric characters also need to be allowed in the pattern?

Comment (IhabAlkabary) [reply src/components/module-scheduler-schedule-dialog.js:234]: Yeah, we need to be conservative here and just remove leading and trailing spaces.

Comment (jguann) [reply src/components/module-scheduler-schedule-dialog.js:234]: Do we also need to allow all special characters then? Commas, apostrophes, colons, etc. might be in the module names as well.

What exactly should we be checking for csv format if alphanumeric + spaces + special characters are allowed?

Comment (IhabAlkabary) [reply src/components/module-scheduler-schedule-dialog.js:234]: I hope there are no commas in the module name, can you check it if this is allowed?
If it's allowed, we need for find a delimiter that is not allowed by Module names.
But this seems not to be part of your PR anymore, can you take that back in your story @Syeda-Azim

Comment (IhabAlkabary) [reply src/components/module-scheduler-schedule-dialog.js:234]: https://d2l.slack.com/archives/C01AU7228U9/p1636576819268900

File: lang/en.js [hunk 6]
```diff
@@ -22,6 +22,11 @@ export default {
 	"scheduleDialog:subject": "Subject Code",
 	"scheduleDialog:moduleIgnoreList": "Module Ignore Strings",
 	"scheduleDialog:deliveryBlocks": "Delivery Blocks",
+	"scheduleDialog:jsonInvalid": "Delivery Blocks must be JSON with valid dates.",
+	"scheduleDialog:jsonInvalidAlert": "Please check your JSON and try again.",
+	"scheduleDialog:moduleIgnoreListInvalid": "Module Ignore Strings must be a comma separated value.",
```
Comment (IhabAlkabary) [inline lang/en.js:27]: lets remove anything to do with Module Ignore List CSV and this PR is good to merge

File: src/components/module-scheduler-schedule-dialog.js [hunk 7]
```diff
@@ -43,6 +44,18 @@ class ScheduleDialog extends LocalizeMixin(LitElement) {
...
 			_readyToShowDialog: {
 				type: Boolean
+			},
+			_missingSubjectOrSessionField: {
+				type: Boolean
+			},
+			_invalidSubjectField: {
+				type: Boolean
+			},
+			_invalidModuleIgnoreStringsField: {
```
Comment (IhabAlkabary) [inline src/components/module-scheduler-schedule-dialog.js:54]: ^

File: src/components/module-scheduler-schedule-dialog.js [hunk 8]
```diff
@@ -141,6 +161,11 @@ class ScheduleDialog extends LocalizeMixin(LitElement) {
 		this.sessionCode = '';
 		this.subjectCode = '';
 		this.moduleIgnoreList = '';
+		this._missingSubjectOrSessionField = false;
+		this._invalidSubjectField = false;
+		this._invalidModuleIgnoreStringsField = false;
```
Comment (IhabAlkabary) [inline src/components/module-scheduler-schedule-dialog.js:166]: ^

File: src/components/module-scheduler-schedule-dialog.js [hunk 9]
```diff
@@ -185,8 +210,38 @@ class ScheduleDialog extends LocalizeMixin(LitElement) {
...
+		if (this.subjectCode) {
+			// Remove all whitespaces
+			this.subjectCode = this.subjectCode.split(' ').join('');
+			if (!this.subjectCode.match(subjectCsvPattern)) {
+				this._invalidSubjectField = true;
+			}
+		}
+
+		// TODO: Update pattern for moduleIgnoreStrings
+		const moduleIgnoreStringsCsvPattern = '^[a-zA-Z0-9]+(,\\s*[a-zA-Z0-9]+)*$';
```
Comment (IhabAlkabary) [inline src/components/module-scheduler-schedule-dialog.js:234]: and here .. I'm not going to point all the places, let's remove the and merge the PR. Thanks.

**Review by IhabAlkabary — APPROVED** [review]

**Pull request merged**